synapse/rest/media/v1/preview_url_resource.py

```diff
@@ -0,0 +1,404 @@
+                    del self.downloads[url]
+                    return media_info
+            media_info = yield download.observe()
+
+            # FIXME: we should probably update our cache now anyway, so that
+            # even if the OG calculation raises, we don't keep hammering on the
+            # remote server.  For now, leave it uncached to aid debugging OG
+            # calculation problems
+
+            logger.debug("got media_info of '%s'" % media_info)
+
+            if self._is_media(media_info['media_type']):
+                dims = yield self._generate_local_thumbnails(
+                    media_info['filesystem_id'], media_info
+                )
+
+                og = {
+                    "og:description": media_info['download_name'],
+                    "og:image": "mxc://%s/%s" % (
+                        self.server_name, media_info['filesystem_id']
+                    ),
+                    "og:image:type": media_info['media_type'],
+                    "matrix:image:size": media_info['media_length'],
+                }
+
+                if dims:
+                    og["og:image:width"] = dims['width']
+                    og["og:image:height"] = dims['height']
+                else:
+                    logger.warn("Couldn't get dims for %s" % url)
+
+                # define our OG response for this media
+            elif self._is_html(media_info['media_type']):
+                # TODO: somehow stop a big HTML tree from exploding synapse's RAM
+
+                try:
+                    tree = html.parse(media_info['filename'])
+                    og = yield self._calc_og(tree, media_info, requester)
+                except UnicodeDecodeError:
+                    # XXX: evil evil bodge
+                    file = open(media_info['filename'])
+                    body = file.read()
+                    file.close()
+                    tree = html.fromstring(body.decode('utf-8', 'ignore'))
+                    og = yield self._calc_og(tree, media_info, requester)
+
+            else:
+                logger.warn("Failed to find any OG data in %s", url)
+                og = {}
+
+            logger.debug("Calculated OG for %s as %s" % (url, og))
+
+            # store OG in ephemeral in-memory cache
+            self.cache[url] = og
+
+            # store OG in history-aware DB cache
+            yield self.store.store_url_cache(
+                url,
+                media_info["response_code"],
+                media_info["etag"],
+                media_info["expires"],
+                json.dumps(og),
+                media_info["filesystem_id"],
+                media_info["created_ts"],
+            )
+
+            respond_with_json_bytes(request, 200, json.dumps(og), send_cors=True)
+        except:
+            # XXX: if we don't explicitly respond here, the request never returns.
+            # isn't this what server.py's wrapper is meant to be doing for us?
```

Comment:
ah, the problem was that i had been `return`ing rather than `raise`ing in the async GET when things went wrong, i think. now fixed.